fetch_db_info.py

```diff
@@ -0,0 +1,843 @@
+    query,
+    function_name,
+    cols,
+    db=None,
+    replicas_port=None,
+    user_db_port=None,
+    user=None,
+    password=None,
+    replicas=True,
+    row_count=500,
+):
+    """
+    Query database (db) and return outputs in chunks.
+
+    :param query: The SQL query to run.
+    :param function_name: The function that was used to collect this data, useful for saving when data is missed due to errors.
+    :param cols: The name of the columns to be used in dataframe for the data collected with SQL.
+    :param db: The database from which the data was collected.
+    :param replicas_port: port for connecting to meta table through ssh tunneling, if used.
+    :param user_db_port: port for connecting to local Sources table through ssh tunneling, if used.
+    :param user: Toolforge username of the tool.
+    :param password: Toolforge password of the tool.
+    :param replicas: False if collecting data from toolsdb user database, True is collecting from other wikimedia databases.
+    :param row_count: Number of rows to get in one query from the database.
+    :return: dataframe
+    """
+
+    offset = 0
+    row_count = row_count
+    max_tries = 3
+
+    while True:
+        retry_counter = 0
+        try:
+            while True:
+                try:
+                    conn = (
+                        db_acc.connect_to_replicas_database(
+                            db, replicas_port, user, password
+                        )
+                        if replicas
+                        else db_acc.connect_to_user_database(
+                            DATABASE_NAME, user_db_port, user, password
+                        )
+                    )
+                    with conn.cursor() as cur:
+                        cur.execute(query + " LIMIT %d OFFSET %d" % (row_count, offset))
+                        df = pd.DataFrame(cur.fetchall(), columns=cols).applymap(
+                            encode_if_necessary
+                        )
+                    close_conn(conn)
+                    break
+                except (
+                    pymysql.err.DatabaseError,
+                    pymysql.err.OperationalError,
+                ) as err:
+                    close_conn(conn)
+                    if retry_counter == max_tries:
+                        raise Exception(err)
+                    print(
+                        "Retrying query of '%s' from %s in 1 minute..."
+                        % (function_name, db)
+                    )
+                    retry_counter += 1
+                    time.sleep(60)
+
+            offset += row_count
+            if len(df) == 0:
+                return
+            yield df
+        except Exception as err:
+            print("Something went wrong. Could not query from %s \n" % db, repr(err))
+            with open("missed_db_info.txt", "a") as file:
+                file.write(function_name + " " + db + "\n")
+            break
+
+
+def save_data(
```

Comment:
`query_data_generator` and `save_data` are great, but I feel like they have more generalized usage and should probably be moved to new file in `utils\`

Reply:
Done in [eacf935](https://github.com/wikimedia/abstract-wikipedia-data-science/pull/22/commits/eacf935035f218b51e84fae8939c81e714744439)